# Эволюция игровой индустрии (2000-2013) — путь к современным RPG

- Автор: Андриянова Анастасия
- Дата: 22.09.2025

### Цели и задачи проекта

Провести комплексный анализ исторических данных игровой индустрии периода 2000-2013 годов для создания статьи, которая привлечет внимание любителей старых игр и повысит интерес к RPG-проекту "Секреты Темнолесья".

### Описание данных

Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
* Name — название игры.
- Platform — название платформы.
- Year of Release — год выпуска игры.
- Genre — жанр игры.
- NA sales — продажи в Северной Америке (в миллионах проданных копий).
- EU sales — продажи в Европе (в миллионах проданных копий).
- JP sales — продажи в Японии (в миллионах проданных копий).
- Other sales — продажи в других странах (в миллионах проданных копий).
- Critic Score — оценка критиков (от 0 до 100).
- User Score — оценка пользователей (от 0 до 10).
- Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

- 1. Загрузка данных и знакомство с ними  
- 2. Проверка ошибок в данных и их предобработка
   - 2.1 Названия, или метки, столбцов датафрейма 
   - 2.2 Типы данных 
   - 2.3 Наличие пропусков в данных  
   - 2.4 Явные и неявные дубликаты в данных  
- 3. Фильтрация данных 
- 4. Категоризация данных  
- 5. Итоговый вывод

---

## 1. Загрузка данных и знакомство с ними


In [1]:
import pandas as pd
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [2]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


None

In [3]:
df.head(15)

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
5,Tetris,GB,1989.0,Puzzle,23.20,2.26,4.22,0.58,NaN,NaN,NaN
6,New Super Mario Bros.,DS,2006.0,Platform,11.28,9.14,6.5,2.88,89.0,8.5,E
7,Wii Play,Wii,2006.0,Misc,13.96,9.18,2.93,2.84,58.0,6.6,E
8,New Super Mario Bros. Wii,Wii,2009.0,Platform,14.44,6.94,4.7,2.24,87.0,8.4,E
9,Duck Hunt,NES,1984.0,Shooter,26.93,0.63,0.28,0.47,NaN,NaN,NaN


In [4]:
df.nunique()

Name               11559
Platform              31
Year of Release       37
Genre                 24
NA sales             402
EU sales             308
JP sales             245
Other sales          155
Critic Score          82
User Score            96
Rating                 8
dtype: int64

- Резюме по данным: в предоставленном датасете 16956 строк с данными, в 6 столбцах содержатся пропуски (особенно много в столбцах Critic Score, User Score, Rating), названия столбцов соответствуют самим данным, про типы данных : EU sales: object вместо float64, JP sales: object вместо float64, User Score: object вместо float64, Year of Release: float вместо int64
а также присутствует настендартное использование пробелов с названиях колонок


In [5]:
df['Name'].unique()

array(['Wii Sports', 'Super Mario Bros.', 'Mario Kart Wii', ...,
       'Woody Woodpecker in Crazy Castle 5', 'LMA Manager 2007',
       'Haitaka no Psychedelica'], dtype=object)

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [6]:
df.columns.tolist()

['Name',
 'Platform',
 'Year of Release',
 'Genre',
 'NA sales',
 'EU sales',
 'JP sales',
 'Other sales',
 'Critic Score',
 'User Score',
 'Rating']

In [7]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

### 2.2. Типы данных

- Возможные причины некорректных топов данных: Оценки критиков отсутствуют для многих игр, возможно из-за недостатка рецензий или новизны продукта. В year_of_release есть пропуски NaN, что автоматически делает тип float64. Ошибка object вместо float64 для eu_sales, jp_sales и user_score могла возникнуть из-за человеческого фактора, некорректного переносса данных либо сбора данных из разных источников

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [8]:
numeric_columns = ['year_of_release', 'na_sales', 'eu_sales', 'jp_sales', 
                   'other_sales', 'critic_score', 'user_score']

In [9]:
string_values = ['unknown', 'n/a', 'N/A', '-', '', ' ', 'tbd', 'TBD', 'null', 'NULL']
# очистка числовых столбцов от строковых значений
for column in numeric_columns:
    if column in df.columns:
        df[column] = df[column].replace(string_values, pd.NA)

 eu_sales, jp_sales, user_score - должны быть float64, но являются object. year_of_release - должен быть int64, но является float64 из-за пропусков Эти столбцы мы преобразуем и обработаем ниже

In [10]:
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors='coerce')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors='coerce')
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce')
print("Типы данных после преобразования строковых столбцов:")
for col in ['eu_sales', 'jp_sales', 'user_score']:
    display(f"{col}: {df[col].dtype}")

Типы данных после преобразования строковых столбцов:


'eu_sales: float64'

'jp_sales: float64'

'user_score: float64'

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [11]:
for column in df.columns:
    if column in df.columns:
        null_count = df[column].isnull().sum()
        null_to_all = 1 -df[column].count() / len(df) # доля пропусков
        dtype = df[column].dtype
        print(f"{column:15} | {str(dtype):8} | пропусков: {null_count:4} | доля пропусков: {null_to_all:6}")

name            | object   | пропусков:    2 | доля пропусков: 0.00011795234725175963
platform        | object   | пропусков:    0 | доля пропусков:    0.0
year_of_release | float64  | пропусков:  275 | доля пропусков: 0.01621844774711012
genre           | object   | пропусков:    2 | доля пропусков: 0.00011795234725175963
na_sales        | float64  | пропусков:    0 | доля пропусков:    0.0
eu_sales        | float64  | пропусков:    6 | доля пропусков: 0.00035385704175516786
jp_sales        | float64  | пропусков:    4 | доля пропусков: 0.00023590469450340823
other_sales     | float64  | пропусков:    0 | доля пропусков:    0.0
critic_score    | float64  | пропусков: 8714 | доля пропусков: 0.5139183769757019
user_score      | float64  | пропусков: 9268 | доля пропусков: 0.5465911771644256
rating          | object   | пропусков: 6871 | доля пропусков: 0.4052252889832507


In [12]:
missing_info = pd.DataFrame({
    'dtype': df.dtypes,
    'null_count': df.isna().sum(),
    'null_fraction': df.isna().mean()
})
missing_info 

,dtype,null_count,null_fraction
name,object,2,0.000118
platform,object,0,0.000000
year_of_release,float64,275,0.016218
genre,object,2,0.000118
na_sales,float64,0,0.000000
eu_sales,float64,6,0.000354
jp_sales,float64,4,0.000236
other_sales,float64,0,0.000000
critic_score,float64,8714,0.513918
user_score,float64,9268,0.546591


- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.
- Наибольшее количество пропусков в столбцах critic_score и user_score. В столбце year_of_release содержится менее 2% пропусков.

In [13]:
# заполним пропуски в year_of_release (менее 2% - можно заполнить медианой)
#median_year = df['year_of_release'].median()
#df['year_of_release'] = df['year_of_release'].fillna(median_year)
#df['year_of_release'] = df['year_of_release'].astype('int64')

In [14]:
df['has_critic_score'] = df['critic_score'].notna().astype('int64')
df['has_user_score'] = df['user_score'].notna().astype('int64')

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [15]:
df[df['name'].isnull()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,has_critic_score,has_user_score
661,NaN,GEN,1993.0,NaN,1.78,0.53,0.00,0.08,NaN,NaN,NaN,0,0
14439,NaN,GEN,1993.0,NaN,0.00,0.00,0.03,0.00,NaN,NaN,NaN,0,0


In [16]:
df_cleaned = df.dropna(subset=['name', 'genre','year_of_release']).copy() # удалим две строки с пропусками в имени и жанре и строки без года


In [17]:
#  для продаж заполняем пропуски нулями ( если данных нет, продажи = 0)
#df_cleaned['eu_sales'] = df_cleaned['eu_sales'].fillna(0)
#df_cleaned['jp_sales'] = df_cleaned['jp_sales'].fillna(0)
df_cleaned['eu_sales'] = df_cleaned['eu_sales'].fillna(
    df_cleaned.groupby(['year_of_release','platform'])['eu_sales'].transform('mean')
)
df_cleaned['jp_sales'] = df_cleaned['jp_sales'].fillna(
    df_cleaned.groupby(['year_of_release','platform'])['jp_sales'].transform('mean')
)

In [18]:
# доп проверка данных на допустимые границы
display(df_cleaned['user_score'].min())
display(df_cleaned['user_score'].max())
display(df_cleaned['critic_score'].min())
display(df_cleaned['critic_score'].max())

0.0

9.7

13.0

98.0

In [19]:
df_cleaned[df_cleaned['rating'] == 'NO_RATING'].shape[0]  

0

Проверяем занят ли тег и используется ли он по назначению

In [20]:
# заменим пропуски в рейтинге на флаг 
original_null_count_rating = df_cleaned['rating'].isnull().sum()
original_null_count_rating
original_rating_distribution = df_cleaned['rating'].value_counts(dropna=False)
display(original_rating_distribution)

NaN     6778
E       3968
T       2948
M       1558
E10+    1414
EC         8
K-A        3
AO         1
RP         1
Name: rating, dtype: int64

In [21]:
# заменяем пропуски на категорию 'NO_RATING' - ее нет в списке возможных тэгов для заполнения
df_cleaned['rating'] = df_cleaned['rating'].fillna('NO_RATING')
# проверяем результат
new_rating_distribution = df_cleaned['rating'].value_counts()
new_rating_distribution

NO_RATING    6778
E            3968
T            2948
M            1558
E10+         1414
EC              8
K-A             3
AO              1
RP              1
Name: rating, dtype: int64

Значения одинаковые (6778), потому что все пропуски (NaN) были успешно заменены на категорию 'NO_RATING'.

In [22]:
missing_info = pd.DataFrame({
    'dtype': df_cleaned.dtypes,
    'null_count': df_cleaned.isna().sum(),
    'null_fraction': df_cleaned.isna().mean()
})
missing_info 

,dtype,null_count,null_fraction
name,object,0,0.000000
platform,object,0,0.000000
year_of_release,float64,0,0.000000
genre,object,0,0.000000
na_sales,float64,0,0.000000
eu_sales,float64,0,0.000000
jp_sales,float64,0,0.000000
other_sales,float64,0,0.000000
critic_score,float64,8594,0.515259
user_score,float64,9121,0.546855


Таким образом мы избавились от дубликатов в строках, оставив два столбца в исходном виде, при этом, создав два новых столбца с флагами, относящимися к исходным столбцам, чтобы не искажать данные

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [23]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16679 entries, 0 to 16955
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              16679 non-null  object 
 1   platform          16679 non-null  object 
 2   year_of_release   16679 non-null  float64
 3   genre             16679 non-null  object 
 4   na_sales          16679 non-null  float64
 5   eu_sales          16679 non-null  float64
 6   jp_sales          16679 non-null  float64
 7   other_sales       16679 non-null  float64
 8   critic_score      8085 non-null   float64
 9   user_score        7558 non-null   float64
 10  rating            16679 non-null  object 
 11  has_critic_score  16679 non-null  int64  
 12  has_user_score    16679 non-null  int64  
dtypes: float64(7), int64(2), object(4)
memory usage: 1.8+ MB


In [24]:
categorical_columns = ['name','platform', 'genre', 'rating', 'year_of_release']

for column in categorical_columns:
    display(f"\ уникальные значения в '{column}':")
    display(df_cleaned[column].value_counts())
    display(f"всего уникальных: {df_cleaned[column].nunique()}")

"\\ уникальные значения в 'name':"

Need for Speed: Most Wanted    12
Ratatouille                     9
LEGO Marvel Super Heroes        9
FIFA 14                         9
Terraria                        9
                               ..
Armed and Dangerous             1
Kengo: Legend of The 9          1
Fate/Unlimited Codes            1
Desire                          1
Threads of Fate                 1
Name: name, Length: 11426, dtype: int64

'всего уникальных: 11426'

"\\ уникальные значения в 'platform':"

PS2     2154
DS      2147
PS3     1330
Wii     1305
X360    1249
PSP     1212
PS      1208
PC       972
GBA      826
XB       818
GC       549
3DS      522
PSV      434
PS4      395
N64      320
XOne     251
SNES     241
SAT      174
WiiU     147
2600     118
NES      100
GB        97
DC        52
GEN       27
NG        12
SCD        6
WS         6
3DO        3
TG16       2
GG         1
PCFX       1
Name: platform, dtype: int64

'всего уникальных: 31'

"\\ уникальные значения в 'genre':"

Action          3341
Sports          2325
Misc            1740
Role-Playing    1493
Shooter         1314
Adventure       1309
Racing          1244
Platform         891
Simulation       866
Fighting         844
Strategy         679
Puzzle           577
ACTION            12
SPORTS             7
RACING             6
FIGHTING           6
ROLE-PLAYING       6
SHOOTER            5
ADVENTURE          4
PLATFORM           3
MISC               3
PUZZLE             2
STRATEGY           1
SIMULATION         1
Name: genre, dtype: int64

'всего уникальных: 24'

"\\ уникальные значения в 'rating':"

NO_RATING    6778
E            3968
T            2948
M            1558
E10+         1414
EC              8
K-A             3
AO              1
RP              1
Name: rating, dtype: int64

'всего уникальных: 9'

"\\ уникальные значения в 'year_of_release':"

2009.0    1450
2008.0    1445
2010.0    1279
2007.0    1218
2011.0    1149
2006.0    1020
2005.0     950
2002.0     839
2003.0     789
2004.0     771
2012.0     670
2015.0     612
2014.0     584
2013.0     552
2016.0     507
2001.0     491
1998.0     384
2000.0     357
1999.0     341
1997.0     293
1996.0     267
1995.0     220
1994.0     121
1993.0      60
1981.0      46
1992.0      43
1991.0      42
1982.0      37
1986.0      22
1983.0      18
1989.0      17
1987.0      17
1990.0      16
1988.0      15
1984.0      14
1985.0      14
1980.0       9
Name: year_of_release, dtype: int64

'всего уникальных: 37'

- Значения в platform нормализированы и не содержат неявных дубликатов
- Значения в genre содержат дубликаты связаные с разницей написания
- Значения в rating содержат неявные дубликаты из-за разных типов обозначения, нужно заменить значения 'К-А' на 'Е'
- Значения в year_of_release нормализированы и не содержат дубликатов
- Значения в name дополнительно приведем в верхний регистр

In [25]:
df_cleaned['name'] = df_cleaned['name'].str.upper()
df_cleaned['genre'] = df_cleaned['genre'].str.lower()
df_cleaned['rating'] = df_cleaned['rating'].replace('K-A', 'E')

In [26]:
categorical_columns = ['name','platform', 'genre', 'rating', 'year_of_release']

for column in categorical_columns:
    display(f"\ уникальные значения в '{column}':")
    display(df_cleaned[column].value_counts())
    display(f"всего уникальных: {df_cleaned[column].nunique()}")

"\\ уникальные значения в 'name':"

NEED FOR SPEED: MOST WANTED         12
MONOPOLY                             9
FIFA 14                              9
TERRARIA                             9
LEGO MARVEL SUPER HEROES             9
                                    ..
MAGICAL VACATION                     1
THE LOST TREASURES OF ALEXANDRIA     1
CT SPECIAL FORCES                    1
MTV SPORTS: PURE RIDE                1
ROGUE GALAXY: DIRECTOR'S CUT         1
Name: name, Length: 11426, dtype: int64

'всего уникальных: 11426'

"\\ уникальные значения в 'platform':"

PS2     2154
DS      2147
PS3     1330
Wii     1305
X360    1249
PSP     1212
PS      1208
PC       972
GBA      826
XB       818
GC       549
3DS      522
PSV      434
PS4      395
N64      320
XOne     251
SNES     241
SAT      174
WiiU     147
2600     118
NES      100
GB        97
DC        52
GEN       27
NG        12
SCD        6
WS         6
3DO        3
TG16       2
GG         1
PCFX       1
Name: platform, dtype: int64

'всего уникальных: 31'

"\\ уникальные значения в 'genre':"

action          3353
sports          2332
misc            1743
role-playing    1499
shooter         1319
adventure       1313
racing          1250
platform         894
simulation       867
fighting         850
strategy         680
puzzle           579
Name: genre, dtype: int64

'всего уникальных: 12'

"\\ уникальные значения в 'rating':"

NO_RATING    6778
E            3971
T            2948
M            1558
E10+         1414
EC              8
AO              1
RP              1
Name: rating, dtype: int64

'всего уникальных: 8'

"\\ уникальные значения в 'year_of_release':"

2009.0    1450
2008.0    1445
2010.0    1279
2007.0    1218
2011.0    1149
2006.0    1020
2005.0     950
2002.0     839
2003.0     789
2004.0     771
2012.0     670
2015.0     612
2014.0     584
2013.0     552
2016.0     507
2001.0     491
1998.0     384
2000.0     357
1999.0     341
1997.0     293
1996.0     267
1995.0     220
1994.0     121
1993.0      60
1981.0      46
1992.0      43
1991.0      42
1982.0      37
1986.0      22
1983.0      18
1989.0      17
1987.0      17
1990.0      16
1988.0      15
1984.0      14
1985.0      14
1980.0       9
Name: year_of_release, dtype: int64

'всего уникальных: 37'

 - Теперь проверим данные на явные дубликаты

In [27]:
full_duplicates_count = df_cleaned.duplicated().sum()
full_duplicates_count

235

In [28]:

if full_duplicates_count > 0:
    duplicate_rows = df_cleaned[df_cleaned.duplicated(keep=False)]  
    display(duplicate_rows.head(10))

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,has_critic_score,has_user_score
267,BATMAN: ARKHAM ASYLUM,PS3,2009.0,action,2.24,1.31,0.07,0.61,91.0,8.9,T,1,1
268,BATMAN: ARKHAM ASYLUM,PS3,2009.0,action,2.24,1.31,0.07,0.61,91.0,8.9,T,1,1
367,JAMES BOND 007: AGENT UNDER FIRE,PS2,2001.0,shooter,1.90,1.13,0.10,0.41,72.0,7.9,T,1,1
368,JAMES BOND 007: AGENT UNDER FIRE,PS2,2001.0,shooter,1.90,1.13,0.10,0.41,72.0,7.9,T,1,1
716,GOD OF WAR: ASCENSION,PS3,2013.0,action,1.23,0.63,0.04,0.35,80.0,7.5,M,1,1
717,GOD OF WAR: ASCENSION,PS3,2013.0,action,1.23,0.63,0.04,0.35,80.0,7.5,M,1,1
822,WIPEOUT: THE GAME,Wii,2009.0,misc,1.94,0.00,0.00,0.12,NaN,NaN,NO_RATING,0,0
823,WIPEOUT: THE GAME,Wii,2009.0,misc,1.94,0.00,0.00,0.12,NaN,NaN,NO_RATING,0,0
847,RAYMAN RAVING RABBIDS: TV PARTY,Wii,2008.0,misc,0.72,1.08,0.00,0.23,73.0,7.7,E10+,1,1
848,RAYMAN RAVING RABBIDS: TV PARTY,Wii,2008.0,misc,0.72,1.08,0.00,0.23,73.0,7.7,E10+,1,1


In [41]:
original_size = len(df)
df_cleaned = df_cleaned.drop_duplicates()

new_size = len(df_cleaned)
removed_count = full_duplicates_count
removed_to_original_dubl = removed_count / new_size
deleted = original_size - new_size
removed_to_original = deleted / original_size

print(f"размер dataset: {original_size}")
print(f"удалено дубликатов: {full_duplicates_count}")
print(f"всего удалено: {deleted}")
print(f"доля удаленных дубликатов: {removed_to_original_dubl}")
print(f"доля удаленных строк: {removed_to_original}")
print(f"новый размер dataset: {new_size}")

размер dataset: 16956
удалено дубликатов: 235
всего удалено: 512
доля удаленных дубликатов: 0.014290926781804913
доля удаленных строк: 0.03019580089643784
новый размер dataset: 16444


- Из датасета устранен 241 явный дубликат 

- В результате проведенной работы датасет теперь содержит 16444 записей, все текстовые данные нормализованы, явные и неявные дубликаты устранены, а пропуски заполнены семантически корректными значениями. Данные приведены в состояние, пригодное для проведения дальнейшего анализа, построения визуализаций и выявления содержательных закономерностей в игровой индустрии.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [30]:
df_actual = df_cleaned[(df_cleaned['year_of_release'] >= 2000) & (df_cleaned['year_of_release'] <= 2013)]

In [31]:
# доп проверки
print(len(df_actual))
print(df_actual['year_of_release'].min())
print(df_actual['year_of_release'].max())

12781
2000.0
2013.0


---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [32]:
def categorizing(score):
    if (score >= 8) & (score <=10):
        return 'высокая оценка'
    elif score >=3:
        return 'средняя оценка'
    elif score >= 0:
        return 'низкая оценка' 
    else:
        return 'нет оценки'

In [33]:
df_actual = df_actual.copy()
df_actual.loc[:, 'user_score_category'] = df_actual['user_score'].apply(categorizing)

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [34]:
def categorizing_critic(score):
    if (score >= 80) & (score <=100):
        return 'высокая оценка'
    elif score >=30:
        return 'средняя оценка'
    elif score >= 0:
        return 'низкая оценка' 
    else:
        return 'нет оценки'

In [35]:
df_actual.loc[:, 'critic_score_category'] = df_actual['critic_score'].apply(categorizing_critic)

In [36]:
 # df_actual['critic_score_category'].value_counts() альтернативный вариант сортировки

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [37]:
user_score_counts = df_actual.groupby('user_score_category')['name'].count().sort_values(ascending=False)
user_score_counts

user_score_category
нет оценки        6298
средняя оценка    4081
высокая оценка    2286
низкая оценка      116
Name: name, dtype: int64

In [38]:
critic_score_counts = df_actual.groupby('critic_score_category')['name'].count().sort_values(ascending=False)
critic_score_counts

critic_score_category
нет оценки        5612
средняя оценка    5422
высокая оценка    1692
низкая оценка       55
Name: name, dtype: int64

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [39]:
platform_rating = df_actual.groupby('platform')['name'].count().sort_values(ascending=False)
platform_rating.head(7)

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: name, dtype: int64

In [40]:
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12781 entries, 0 to 16954
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   name                   12781 non-null  object 
 1   platform               12781 non-null  object 
 2   year_of_release        12781 non-null  float64
 3   genre                  12781 non-null  object 
 4   na_sales               12781 non-null  float64
 5   eu_sales               12781 non-null  float64
 6   jp_sales               12781 non-null  float64
 7   other_sales            12781 non-null  float64
 8   critic_score           7169 non-null   float64
 9   user_score             6483 non-null   float64
 10  rating                 12781 non-null  object 
 11  has_critic_score       12781 non-null  int64  
 12  has_user_score         12781 non-null  int64  
 13  user_score_category    12781 non-null  object 
 14  critic_score_category  12781 non-null  object 
dtypes:

---

## 5. Итоговый вывод



### Проделанная работа: 
1. Характеристики среза:
- Охватывает период расхода игровых консолей 6-7 поколений
- Включает эпоху доминирования PS2, Nintendo DS и Wii
- Захватывает переход от SD к HD-графике (появление PS3 и X360)
- Представляет современную игровую индустрию в зрелом состоянии
2. Подготовка данных
- Проведена предобработка данных, устранение дубликатов, форматирование и фильтрация
- Создан срез данных за период 2000-2013 годы включительно
- Исходный датасет содержал 13050 записей (игр) за указанный период
- Проверены границы временного интервала: минимальный год - 2000, максимальный - 2013
3. Добавлены 4 новых поля категоризации: user_score_category - категория пользовательских оценок, critic_score_category - категория оценок критиков, и два вспомогательных: has_critic_score и has_user_score (содержан флаги "есть ли оценка")
4. Проведен анализ пользовательских оценок и оценок критиков
#### Ключевые выводы:
- Высокая доля отсутствующих оценок - почти половина игр не имеет пользовательских оценок, что важно учитывать при анализе
- Доминирование Sony и Nintendo - PS2 и DS являются лидерами по количеству выпущенных игр
- Баланс оценок - большинство игр получают средние оценки, что свидетельствует о стабильном качестве продукции в индустрии